feat(commander): wire current-residual motor-failure detector + offline replay tool (#27917)
PX4/PX4-Autopilot: docs/en/debug/motor_failure_replay.md

# Offline Motor Failure Replay

<Badge type="tip" text="main (PX4 v2.0)" />

Two command line tools calibrate the [motor failure detector](../config/motor_failure_detection.md) against a vehicle's own flight logs.
`mfd_fit` derives the current model from a set of logs, and `mfd_replay` runs the real detector over a log and prints what it would have decided for each motor.

They exist because the [threshold parameters](../config/motor_failure_detection.md#choosing-parameter-values) (`MOTFAIL_*`) are specific to an airframe, so the only place to get them is that airframe's own flights.
`mfd_replay` links the detector library itself, so a replay runs the same code that runs in flight.

> One flight is rarely enough to characterise a vehicle, so most of the work below runs over a directory of them.

Reviewer: ```suggestion
One flight is rarely enough to characterise a vehicle, so usually we run the tools over a directory of logs.
```

> Typical uses:

Reviewer: ```suggestion
The tools are used to:
```

- Confirm a candidate configuration never trips on a set of healthy flights.
- Find the tightest thresholds a vehicle's data supports, by sweeping values and re-running.
- Check after the fact whether the detector would have flagged a motor in a flight where something went wrong.

## Building

> To build the tool:

Reviewer: ```suggestion
To build the tools:
```

1. Build PX4 SITL (to populate the build directory)

   ```bash
   make px4_sitl_default # once, if the build directory does not exist yet
   ```

2. Build the tool binaries:

   ```bash
   cmake -DMFD_REPLAY_TOOL=ON build/px4_sitl_default
   ninja -C build/px4_sitl_default mfd_replay mfd_fit
   ```

This produces `build/px4_sitl_default/mfd_replay` and `build/px4_sitl_default/mfd_fit`.
`mfd_fit` derives the current model from logs; `mfd_replay` then runs the detector with it.

## Fitting the Current Model

> 

Reviewer: What does "fitting" mean, and is this the first thing you do? 

As a reader this is the first thing I see after building, so I would assume you'r telling me this is what I do next. A sentence explaining why you fit the current model and what the result after this step helps people understand the rest.

SO a sentence like "The first thing is to do X, as this results in Y, Z, that you can then do A with. .. (or similar)

Where X is "fit the model" to "hint at the next step". 

To put it another way, after reading this section I wasn't sure what the outputs were I was after, or why. I am fairly sure I could unpick this by reading the motor failure detection in parallel, but I shouldn't have to - the instructions should proceed though the steps clearly and as turnkey as possible.

The initial sentence provides the context of what follows. If you reread it at the end of the section you should be able to say "yeah, that's what I put in, got out, and why I bothered"

`mfd_fit` reads ESC current and motor command from as many logs as you pass it, and fits the line the detector uses.
It counts only armed samples where the ESC reported a current, and skips motors that were switched off or running in reverse.

```bash
./build/px4_sitl_default/mfd_fit ~/logs/healthy/*.ulg
```

```text
# 3 logs, 80697 armed samples
# I_expected = 6.91*u + 0.37 A   (residual sigma 0.50 A)
# mean command 0.387, mean current 3.04 A

MOTFAIL_C2T 6.91
MOTFAIL_IDLE 0.37

# per-motor mean residual against this model [A]
#   motor 1: -0.12  (sigma 0.64, n 13484)
#   ...
#   motor 6: -0.10  (sigma 0.46, n 13360)
# worst per-motor bias 0.12 A -- the trip bands have to clear this before any fault.
```

> Set what the tool prints.

Reviewer: Is this an instruction? I am not sure what is intended in ths paragraph. Is it "See what the tool prints above"?

Reviewer: I think it is. I'd say something like "The text above shows a typical log output. This prints the parameters that best fit the model (the "candidate parameters"), the XX, the YYY."

Then go into the detail.

> On some vehicles the fit puts the offset below zero, and [MOTFAIL_IDLE](../config/motor_failure_detection.md#MOTFAIL_IDLE) cannot be set below 0: use 0 in that case.

Reviewer: If this offset refers to the text, how is that labeled in the text?

My guess is that in `motor 1: -0.12  (sigma 0.64, n 13484)` maybe it is -0.12? We need to be precise.

Nothing is lost by it, because the bands are then chosen against the residual that results, so they absorb the difference.

Two numbers matter beyond the parameters themselves:

- **residual sigma**: the spread the trip bands have to clear, which caps how tight they can be.
- **worst per-motor bias**: what the shared model costs the worst motor. The detector has one model for all motors, so this is error the bands absorb before any fault does.

> Pass the whole set in one command: the fit is pooled over every log, and a hundred logs takes under a second.

Reviewer: I would probably put this up the top of the section too.

::: details Running it over a large set of logs
`mfd_fit` takes the logs themselves rather than a directory, so use a glob, and a recursive one (`**/*.ulg`) if they are nested.

Memory does not grow with the size of a log or with how many you pass (91 logs and 1.5 GB fit in 0.8 s and 5 MB), so there is never a reason to split the work up.
In particular, do not pipe the list through `xargs`: it starts a new process once the list gets long, which would give you several partial fits instead of one.

Check the `# N logs` line against the number of files you passed.
A log without ESC current contributes nothing and is named on stderr, which is easy to miss on a large set, so redirect it with `2> skipped.txt` if you want to know which ones dropped out.
:::

::: warning
Fit on healthy flights only.
The fit cannot tell a faulty motor from a healthy one, so a single failure-test log pulls the model toward that fault and shifts the residual of every healthy motor with it.
:::

Fit the offset, do not assume one.
A borrowed offset biases every healthy motor's residual by that amount, which the undercurrent band then has to absorb.

> Taking another airframe's 2.88 A instead of the fitted 0.37 A turns all six healthy motors into failures, at bands that were clean:

Reviewer: I'm not sure what this means. I think the whole section structure feels like it doesn't flow.
- What you do at high level - run over a set of known good logs, as many as you like
- The data is X, Y, Z.
- Do this with it.

| `motor`   | Motor index, counting from 0, as mapped from each ESC's actuator function.                                                                           |
| `min[A]`  | Deepest the filtered residual went below the model, over the whole flight and every sampling phase. This is what the undercurrent band has to clear. |
| `max[A]`  | Highest it went above the model, likewise. This is what the overcurrent band has to clear.                                                           |
| `trips`   | How many of the sampling phases flagged this motor.                                                                                                  |
| `verdict` | `ok`, `FAILED (all phases)`, or `FAILED (sampling-dependent!)` when only some phases flagged it.                                                     |

Motors are numbered from 1, the same as in the `Motor 1 undercurrent detected` message the vehicle reports.

Exit codes make the tool usable as a check in a script: `0` if every motor stayed healthy, `1` if any motor was flagged, `2` if the log could not be used.

> ### The Sampling Phase Sweep

Reviewer: First sentence set the scene of why this section is important, and why at this point.

"The signally phase sweep is run when XXX [during flight?], in order to YYY"

On the vehicle the check runs at 10 Hz, so it sees only a subset of the ESC samples, and _which_ subset depends on the alignment between the check and the telemetry stream.
That alignment cannot be reconstructed from a log.

The tool therefore runs several detectors over the same log, each starting its 10 Hz grid at a different offset, and reports the spread.
A single-phase answer would be a lower bound on the trip risk, not a verdict.

Read the result as follows:

- `0/5`: the configuration did not trip on this flight, and `min`/`max` say by how much it cleared each band.
- `0/5` but with a note printed below the table: one of the extremes moved by more than 30% between phases, so this run happened not to trip and the same flight could trip on the vehicle.
- `1/5` … `4/5` (`sampling-dependent!`): treat this as a false positive, not as a pass.
- `5/5`: it would have tripped.

### Where the Configuration Comes From

> 

Reviewer: Same comment. "The configuration you see in your XXX, that is used to YYYY" ... then the rest.

1. `--set NAME=value` on the command line.
2. The value recorded in the log, if the log contains the `MOTFAIL_*` parameters.
3. A built-in fallback.

The header line always prints the configuration that was used, and names any parameter that came from `--set`.

::: warning
The built-in fallback is an example fit for one particular vehicle, not a recommended default.
On another airframe it will report failures on perfectly healthy flights.
Always replay with the configuration you are actually evaluating.
:::

Overriding one knob at a time is how the band sweeps below are done:

```bash
./build/px4_sitl_default/mfd_replay flight.ulg --set MOTFAIL_UNDER=1.75 --set MOTFAIL_UND_TIME=0.7
```

The accepted names are [MOTFAIL_C2T](../config/motor_failure_detection.md#MOTFAIL_C2T), [MOTFAIL_IDLE](../config/motor_failure_detection.md#MOTFAIL_IDLE), [MOTFAIL_UNDER](../config/motor_failure_detection.md#MOTFAIL_UNDER), [MOTFAIL_UND_TIME](../config/motor_failure_detection.md#MOTFAIL_UND_TIME), [MOTFAIL_OVER](../config/motor_failure_detection.md#MOTFAIL_OVER) and [MOTFAIL_OVR_TIME](../config/motor_failure_detection.md#MOTFAIL_OVR_TIME).

## Choosing the Bands

> 

Reviewer: How did I get here in the flow. 

"After running XXX, you will have YYYY, which you need to use to chose the bands"
